### PR TITLE
Multiple OST daemon 

### PR DESCRIPTION
**Pull request by brtknr**

### CODE REVIEW

File: tasks/tuning.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      echo always > /sys/kernel/mm/transparent_hugepage/defrag
+      for dev in {{ beegfs_block_devices | join(' ') }}
+      do
+        echo mq-deadline > /sys/block/${dev}/queue/scheduler
+        echo 2048 > /sys/block/${dev}/queue/nr_requests
+        echo 4096 > /sys/block/${dev}/queue/read_ahead_kb
+        echo 256 > /sys/block/${dev}/queue/max_sectors_kb
+      done
+  become: yes
+  notify: Restart rc-local
```
[the remark this answers is not in the document]
Comment (markgoddard) [reply tasks/tuning.yml]: Alternative approach that avoids this question is to write out a script, execute the script from rc.local and replace 'Restart rc-local' with 'Tune storage server'.

File: defaults/main.yml [hunk 2]
```diff
@@ -1,5 +1,9 @@
 ---
 beegfs_repo_url: "https://www.beegfs.io/release/latest-stable/dists/beegfs-rhel7.repo"
+beegfs_block_devices: []
+beegfs_path_oss: "{{ (['/data/beegfs/beegfs_oss/'] * beegfs_block_devices|length) | zip(beegfs_block_devices) | map('join') | list }}"
+beegfs_dev_oss: "{{ (['/dev/'] * beegfs_block_devices|length) | zip(beegfs_block_devices) | map('join') | list }}"
+beegfs_port_oss: []
```
[the remark this answers is not in the document]
Comment (oneswig) [reply defaults/main.yml]: Having a list of dicts for OSS devices would make sense, provided we could apply defaults without introducing excessive boilerplate replication.  Importing user parameters and inserting defaults to an internally-managed list of dicts, for example.

File: handlers/main.yml [hunk 3]
```diff
@@ -48,4 +49,10 @@
   delay: 3
   register: result
   until: result is success
+
+- name: Restart rc-local
+  service:
+    name: rc-local
+    state: restarted
+  become: yes
```
Comment (oneswig) [inline handlers/main.yml]: Here be dragons - what if rc.local had other things in it too (as it easily could).  Perhaps defining a systemd unit for those tuning parameters would be safer.

File: tasks/oss.yml [hunk 4]
```diff
@@ -10,17 +10,39 @@
   set_fact:
     mounted_devs: "{{ ansible_mounts | map(attribute='device') | list }}"
 
-- name: Ensures BeeGFS storage service is not running
+- name: Ensures BeeGFS client service is not running
```
[the remark this answers is not in the document]
Comment (brtknr) [reply tasks/oss.yml]: Yes unfortunately otherwise it never unmounts

File: tasks/create.yml [hunk 5]
```diff
@@ -52,6 +52,8 @@
 - include: client.yml
   when: beegfs_enable.client | bool and beegfs_path_client not in ansible_mounts
 
+- include: tuning.yml
```
[the remark this answers is not in the document]
Comment (brtknr) [reply tasks/create.yml]: Yes, I think so...

File: tasks/oss.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (markgoddard) [reply tasks/oss.yml]: Maybe the clients should all be stopped first in clients.yml?

[the remark this answers is not in the document]
Comment (brtknr) [reply tasks/oss.yml]: I moved it to create.yml. It cannot be in clients.yml because bulk of it has to run after everything else is ready.

**Comment by brtknr** [comment]
This is ready for review. I have also updated p3-appliances and hpcdirect-appliances to reflect the changes here:
- https://github.com/stackhpc/p3-appliances/pull/35
- https://github.com/VerneGlobal/hpcdirect-appliances/pull/3

File: handlers/main.yml (hunk 3 above)
Comment (brtknr) [reply handlers/main.yml]: I have moved these jobs over to systemd unit files.

File: handlers/main.yml [hunk 6]
```diff
@@ -15,9 +15,10 @@
 
 - name: Restart BeeGFS storage service
   service:
-    name: beegfs-storage
+    name: "beegfs-storage@{{ item }}"
     state: restarted
   become: yes
+  loop: "{{ beegfs_block_devices }}"
   when: beegfs_enable.oss | bool
   notify: Ensure that BeeGFS storage service has really started
```
[the remark this answers is not in the document]
Comment (brtknr) [reply handlers/main.yml]: Yup. Although `beegfs_block_devices` doesn't exist anymore, so that was a typo.

File: tasks/oss.yml [hunk 7]
```diff
@@ -10,60 +10,114 @@
...
+    remote_src: yes
+    force: no
+  loop: "{{ beegfs_oss }}"
+  become: yes
+  notify: Restart BeeGFS storage service
+
 - name: Prepare storage devices
   block:
   - name: Unmount storage device if beegfs_force_format is true
+    vars:
```
Comment (markgoddard) [inline tasks/oss.yml]: Move these vars to the block scope.

File: tasks/destroy.yml [hunk 8]
```diff
@@ -17,37 +46,41 @@
...
-  - { name: beegfs-storage, enable: "{{ beegfs_enable.oss }}" }
   - { name: beegfs-admon, enable: "{{ beegfs_enable.admon }}" }
-  - { name: beegfs-client, enable: "{{ beegfs_enable.client }}" }
-  - { name: beegfs-helperd, enable: "{{ beegfs_enable.client }}" }
+  - { name: beegfs-tuning, enable: "{{ beegfs_enable.oss }}" }
 
 - name: Clear storeStorageDirectory
   lineinfile:
-    path: /etc/beegfs/beegfs-storage.conf
+    path: "/etc/beegfs/{{ item.dev }}.d/beegfs-storage.conf"
```
[the remark this answers is not in the document]
Comment (brtknr) [reply tasks/destroy.yml]: Yep

File: tasks/oss.yml (hunk 7 above)
Comment (brtknr) [reply tasks/oss.yml]: They are using items from a loop...

Comment (markgoddard) [reply tasks/oss.yml]: :+1:

File: files/beegfs-tuning.service [hunk 9]
```diff
@@ -0,0 +1,13 @@
+[Unit]
+Description=BeeGFS Tuning Service
+Documentation=http://www.beegfs.com/content/documentation/
+Requires=network-online.target
+After=network-online.target
+
+[Service]
+ExecStart=/bin/bash /opt/beegfs/beegfs-tuning.sh
```
Comment (oneswig) [inline files/beegfs-tuning.service]: On a converged system, we might want to call this beegfs-oss-tuning - it's not about metadata or client, right?

File: defaults/main.yml [hunk 10]
```diff
@@ -1,45 +1,53 @@
...
+beegfs_mgmt_host: # "{{ groups['cluster_beegfs_mgmt'] | first }}"
+
+# Metadata storage options
+beegfs_meta_path: "/data/beegfs/beegfs_meta"
+beegfs_meta_dev: # /dev/sdb
+beegfs_meta_fstype: # "ext4"
 
 # Client mount configs
-beegfs_client_confs: []
+beegfs_client:
```
Comment (oneswig) [inline defaults/main.yml]: Is this a good default value for beegfs_client - what about an empty list (but the dict structure still documented in comments)

File: README.md [hunk 11]
```diff
@@ -51,20 +51,22 @@ And a corresponding playbook as this (`beegfs.yml`):
           meta: "{{ inventory_hostname in groups['cluster_beegfs_mds'] }}"
```
Comment (markgoddard) [inline README.md]: The README could do with a little filling out, especially for non-scalar variables. That can be done in a later commit though.

File: tasks/client.yml [hunk 12]
```diff
@@ -62,46 +62,56 @@
   notify: Restart BeeGFS client service
 
 - name: Make of copy of BeeGFS client config file if it doesn't exist
+  vars:
+    mgmt_host: "{{ item.mgmt_host | default(beegfs_mgmt_host) }}"
```
Comment (oneswig) [inline tasks/client.yml]: Would it work if you iterated over the include of a file of tasks for mounting one client?  That might save a lot of replication of defining mgmt_host and repeating with_items.

File: tasks/client.yml [hunk 13]
```diff
@@ -62,46 +62,56 @@
...
   become: yes
 
 - name: Copy default beegfs-client config file to enable beegfs_ctl in the management host
+  vars:
+    mgmt_host: "{{ item.mgmt_host | default(beegfs_mgmt_host) }}"
   copy:
     remote_src: yes
     dest: /etc/beegfs/beegfs-client.conf
-    src: "/etc/beegfs/beegfs-client-{{ item.beegfs_host_mgmt }}.conf"
+    src: "/etc/beegfs/beegfs-client-{{ mgmt_host }}.conf"
```
Comment (markgoddard) [inline tasks/client.yml]: This probably isn't going to work with multiple clients - they'll all overwrite each other.

**Pull request merged**